iree/vm/list.c

```diff
@@ -686,3 +686,124 @@ iree_status_t iree_vm_list_register_types(void) {
+  do {                 \
+    typeof(a) t = a;   \
+    a = b;             \
+    b = t;             \
+  } while (0);
+
+static bool iree_vm_list_is_the_same_type(const iree_vm_list_t* list_a,
+                                          const iree_vm_list_t* list_b) {
+  return list_a->element_type.value_type == list_b->element_type.value_type &&
+         list_a->element_type.ref_type == list_b->element_type.ref_type;
+}
+
+static bool iree_vm_list_is_compatiable_allocator(
+    const iree_vm_list_t* list_a, const iree_vm_list_t* list_b) {
+  // Reject if both lists are allocated on the stack. The allocator is
+  // iree_allocator_null if the list is allocated on the stack.
+  //
+  // benvanik: swapping between stack allocated lists is probably fine, though
+  // to be conservative we could just fail.
+  if (iree_allocator_is_null(list_a->allocator) ||
+      iree_allocator_is_null(list_b->allocator))
+    return false;
+
+  return list_a->allocator.ctl == list_b->allocator.ctl;
+}
+
+iree_status_t iree_vm_list_swap(iree_vm_list_t* list_a,
+                                iree_vm_list_t* list_b) {
+  if (!iree_vm_list_is_the_same_type(list_a, list_b))
+    return iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                            "try to swap two different types of lists.");
+
+  if (!iree_vm_list_is_compatiable_allocator(list_a, list_b))
+    return iree_make_status(
+        IREE_STATUS_FAILED_PRECONDITION,
+        "try to swap two incompatible allocators of lists.");
+
+  // FIXME: can we exchange ref_object?
+  std_swap(list_a->ref_object, list_b->ref_object);
+  // FIXME: can we exchange allocator.self?
+  std_swap(list_a->allocator.self, list_b->allocator.self);
+
+  std_swap(list_a->capacity, list_b->capacity);
+  std_swap(list_a->count, list_b->count);
+  std_swap(list_a->element_size, list_b->element_size);
+  std_swap(list_a->storage_mode, list_b->storage_mode);
+  std_swap(list_a->storage, list_b->storage);
+
+  return iree_ok_status();
+}
+
+iree_status_t iree_vm_list_copy(const iree_vm_list_t* src_list,
+                                iree_host_size_t src_offset,
+                                iree_vm_list_t* dst_list,
+                                iree_host_size_t dst_offset,
+                                iree_host_size_t count) {
+  if (!iree_vm_list_is_the_same_type(src_list, dst_list))
+    return iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                            "try to copy two different types of lists.");
+
+  if (src_offset + count > src_list->count) {
+    return iree_make_status(
+        IREE_STATUS_OUT_OF_RANGE,
+        "try to copy %zu elements with offset %zu which is out of bounds (%zu)",
+        count, src_offset, src_list->count);
+  }
+
+  if (dst_offset + count > dst_list->count) {
+    return iree_make_status(
+        IREE_STATUS_OUT_OF_RANGE,
+        "try to copy %zu elements with offset %zu which is out of bounds (%zu)",
+        count, dst_offset, dst_list->count);
+  }
+
```

Comment:
since all storage offsets are calculatable without the storage mode we can hoist those up here:
```c
uint8_t* src_ptr = src_list->value_storage + src_offset + src_list->element_size;
uint8_t* dst_ptr = dst_list->value_storage + dst_offset + dst_list->element_size;
```